gtsam/linear/JacobianFactor.cpp

```diff
@@ -560,8 +560,9 @@ void JacobianFactor::hessianDiagonalAdd(VectorValues& d) const {
         model_->whitenInPlace(column_k);
       dj(k) = dot(column_k, column_k);
     }
-    if(d.exists(j)) {
-      d.at(j) += dj;
+    auto item = d.find(j);
```

Comment:
So, I think we can do *even* better! Currently, we do a find (one traversal) and another if it's not there (emplace). By *always* doing emplace and inspecting its return value we only do *one* traversal per key:

```
If the function successfully inserts the element (because no equivalent element existed already in the map), the function returns a pair of an iterator to the newly inserted element and a value of true.
Otherwise, it returns an iterator to the equivalent element within the container and a value of false.
```

Reply:
```c++
  /* ************************************************************************* */
  VectorValues::iterator VectorValues::emplace(Key j, const Vector& value) {
#ifdef TBB_GREATER_EQUAL_2020
    std::pair<iterator, bool> result = values_.emplace(j, value);
#else
    std::pair<iterator, bool> result = values_.insert(std::make_pair(j, value));
#endif
    if(!result.second)
      throw std::invalid_argument(
      "Requested to emplace variable '" + DefaultKeyFormatter(j)
      + "' already in this VectorValues.");
    return result.first;
  }
```

This will throw if value exists. Should I make a function to access the inner `values_`?

Reply:
Oh, dang! I had no idea the semantics of our emplace were different. That’s actually terrible :-) I think you should change this to an in-line straight call to emplace in the header, and just return the result. Might have to check current uses of emplace but I suspect none of them actually use the return value.

Reply:
I think here the aim is to avoid the allocation of `Vector dj(nj)`? If so then the emplace will need a new Vector, thus resummoning the allocation?

Reply:
Yes - either use the memory already *found* by emplace in the tree, or use the newly allocated memory it just created (if the key was not in the tree yet). Either way, emplace will give you a reference to the memory, so no allocation should be necessary in your code. *and* the tree is traversed only once.

Reply:
"If the function successfully inserts the element" this element still needs to be allocated first:
```
Vector dj(nj);

if(vv.emplace(j, dj)) {
  ...
} else {
  ...
}
```

Reply:
That's interesting. This is the copy-constructor of a Vector, right? But there is also one that takes a size_t and just allocates?

Reply:
Maybe this needs Arg forwarding - so *all* constructors for Vector are automatically available?

Reply:
That is the `size_t` constructor, but this is still an allocation. Wonder if there is something that finds and allocates if not found...

Reply:
There is `try_emplace`, but only with C++17/////////

Reply:
I don't think you need it. emplace will forward arguments, right. So, for normal maps this should work:

```
size_t nj = ...
auto item = emplace(j, nj);
auto& dj = *item.first;
if (item.second) dj.setZero();
for () {
  dj(k) += ...
}
```

Try and work it out entirely before sending next comment?

Reply:
Turns out the way we are using `emplace` is probably wrong. The argument to `emplace` is the argument to the object constructor, so basically we are calling the copy constructor all the time when we have that `VectorValues::emplace` indirection :\

Reply:
Exactly

Reply:
https://stackoverflow.com/questions/31804025/why-is-unordered-map-find-insert-faster-than-insert-check-for-success

I think as we are on C++11 it's still better to find and insert?

Reply:
done